docs/proposals/20191030-machine-health-checking.md

```diff
@@ -105,14 +104,17 @@ As an operator of a Management Cluster, I want my machines to be self-healing an
 
 ### Implementation Details/Notes/Constraints
 
-#### Machine annotation:
+#### Machine conditions:
 ```go
-const MachineUnhealthyAnnotation = "machine.cluster.x-k8s.io/unhealthy"
+const ConditionHealthCheckSucceeded ConditionType = "HealthCheckSucceeded"
+const ConditionOwnerRemediationSucceeded ConditionType = "OwnerRemediationSucceeded"
 ```
 
-This annotation is applied by the MHC controller and it is expected that it
-will never be removed. The only current mediation strategy is deletion, which
-is to be handled by the machine's owning controller.
+Both of these conditions are applied by the MHC controller. ConditionHealthCheckSucceeded should only be updated by the MHC after running the health check.
+If a health check passes after it has failed, it should remove the ConditionOwnerRemediationSucceeded condition and set ConditionHealthCheckSucceeded to True.
+ConditionOwnerRemediationSucceeded is set to False after a health check fails, but should be changed to True by the owning controller after remediation succeeds.
+If remediation fails ConditionOwnerRemediationSucceeded can be updated to a higher severity and the reason can be updated to aid in troubleshooting.
+In the event that ConditionOwnerRemediationSucceeded is still False when the next health check is run, the MHC controller will not update this condition.
```

Comment:
Hmm, this one is stretching my brain a little. When we increase the conditions we increase the number of dimensions of weird state combinations we can get. Not even including the possibility of some conditions being unknown or not present.

h = healthcheck succeeded True
n = remediation needed True
s = remediation succeeded True
H = healthcheck succeeded False
N = remediation needed False
S = remediation succeeded False

with 2 conditions:
- Hs = good, "happy path"
- HS = a failure to remediate, but we can retry or leave for manual intervention
- hs = good, this will be temporarily the state when the deletion of the machine is pending or if/when we support non-deletion remediation
- hS = undefined, if remediation fails but health check is needed it's not clear how we got here. Maybe subsequent healthchecks passed for *reasons*, or the healthcheck was changed? Same concerns as raised in your comment [here](https://github.com/kubernetes-sigs/cluster-api/pull/3056/files#r426709068).

with 3 conditions:
- Hns = good, same as Hs
- HnS = good, same as HS
- HNs = undefined, failing healthcheck, don't need remediation, remediation succeeded
- HNS = undefined, failing healthcheck, don't need remediation, remediation failed
- hns = good, same as hs
- hNs = undefined, passing healthcheck, don't need remediation, remediation succeeded
- hNS = undefined, passing healthcheck, don't need remediation, remediation succeeded
- hnS = undefined, passing healthcheck, need remediation, remediation succeeded